neuralpp/symbolic/interval.py

```diff
@@ -0,0 +1,190 @@
+        return self.lower_bound.internal_object_eq(other.lower_bound) and \
+               self.upper_bound.internal_object_eq(other.upper_bound)
+
+    def __iter__(self) -> Iterable[Int]:
+        """
+        If upper and lower bounds are constant, return a range that's iterable.
+        Otherwise, raise
+        """
+        match self.lower_bound, self.upper_bound:
+            case Constant(value=l, type=builtins.int), Constant(value=r, type=builtins.int):
+                return range(l, r)
+            case _:
+                raise TypeError("Lower and upper bounds must both be Constants!")
+
+    @property
+    def size(self) -> Expression:
+        return self.upper_bound - self.lower_bound + 1
+
+    def to_context(self, index: Variable) -> Context:
+        result = Z3SolverExpression() & index >= self.lower_bound & index <= self.upper_bound
+        assert isinstance(result, Context)  # otherwise lower_bound <= upper_bound is unsatisfiable
+        return result
+
+
+class DottedIntervals(BasicExpression):
+    def __init__(self, interval: ClosedInterval, dots: List[Expression]):
+        super().__init__(Set)
+        self._interval = interval
+        self._dots = dots
+
+    @property
+    def interval(self) -> ClosedInterval:
+        return self._interval
+
+    @property
+    def dots(self) -> List[Expression]:
+        return self._dots
+
+    @property
+    def subexpressions(self) -> List[Expression]:
+        return [self.interval] + self.dots
+
+    def set(self, i: int, new_expression: Expression) -> Expression:
+        raise NotImplementedError("TODO")
+
+    def replace(self, from_expression: Expression, to_expression: Expression) -> Expression:
+        raise NotImplementedError("TODO")
+
+    def internal_object_eq(self, other) -> bool:
+        raise NotImplementedError("TODO")
+
+    @property
+    def __iter__(self) -> Iterable[Constant]:
+        raise NotImplementedError("TODO")
+
+
+def from_constraints(index: Variable, constraint: Context) -> Expression:
+    """
+    @param index:
+    @param constraint:
+    @return: an if-then-else tree whose leaves are DottedIntervals
+
+    This currently only supports the most basic of constraints
+    For example, x > 0 and x <= 5
+    This should return an interval [1, 5]
+    More complicated cases will be added later
+    """
+    closed_interval = ClosedInterval(None, None)
+    exceptions = []
+    for subexpression in constraint.subexpressions:
+        if (isinstance(subexpression, FunctionApplication)):
+            closed_interval, exceptions = _extract_bound_from_constraint(index, subexpression, closed_interval, exceptions)
+
+    return DottedIntervals(closed_interval, exceptions)
+
+def _extract_bound_from_constraint(
+    index: Variable,
+    constraint: Expression,
+    closed_interval: ClosedInterval,
+    exceptions: List[Expression]
+) -> Tuple(ClosedInterval, List[Expression]):
+    """
+    Helper function for from_constraints
```

Comment:
I don't think a reader of this docstring will be able to completely understand what the function is doing. They would have to read the code where it is used and figure it out, which would take time.
Can you please refine it? I think the main thing to include is the relationship between output and inputs. The function returns a closed_interval and exceptions; what holds between that and the input closed_interval and exceptions, for example?